### PR TITLE
Initial commit for versioned template (.tf) files to create GKE clusers for OpenEBS-Staging

### LINKED ISSUE #1199: Setup an OpenEBS staging environment on the cloud
## Is this a BUG REPORT or FEATURE REQUEST?
FEATURE REQUEST

**What happened**:

(Refer parent issue #1198 ) 

Create a staging environment on the cloud (preferably GKE) into which regular updates can be made, thereby providing insights into "real-world" issues while using OpenEBS. Ensure this setup : 

- Has multiple applications w/ load generation (ranging from idle/intermittent-to-frequent) 
- Multiple namespaces
- Different users 
- Logging & Monitoring enabled

### LINKED ISSUE #1200: Create templates (version controlled IaC) for OpenEBS staging cluster
## Is this a BUG REPORT or FEATURE REQUEST?
FEATURE REQUEST

**What happened**:

(Refer parent issue #1198 )

Create templates (preferably terraform .tf files)  to build the OpenEBS staging GKE cluster while also making cluster modifications via these to ensure an identical infrastructure can be brought up w/o manual intervention.

**Anything else we need to know?**:

Some good references include : 

- https://www.terraform.io/docs/providers/google/r/container_cluster.html
- https://github.com/l337ch/terraform-gke
- https://www.pelo.tech/blog/running-kubernetes

### PR DESCRIPTION
**Pull request by ksatchit**
**What this PR does / why we need it**:

- This PR provides the ability to maintain GKE cluster configurations as versioned terraform template files 

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: 

fixes #1199 
fixes #1200 

**Special notes for your reviewer**:

- Needs the service account key of compute-engine-developer 
- Creates cluster in default VPC. Doesn't include add-ons like node/pod autoscale 

Both these and few other  can be added via terraform module files if/when needed.

### CODE REVIEW

**Comment by ksatchit** [comment]
The terraform O/P commands are provided here : 

```
ciuser@OpenEBSClient:~/oebs/gke$ terraform init

Initializing provider plugins...
- Checking for available provider plugins on https://releases.hashicorp.com...
- Downloading plugin for provider "google" (1.6.0)...

The following providers do not have any version constraints in configuration,
so the latest version was installed.

To prevent automatic upgrades to new major versions that may contain breaking
changes, it is recommended to add version = "..." constraints to the
corresponding provider blocks in configuration, with the constraint strings
suggested below.

* provider.google: version = "~> 1.6"

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
ciuser@OpenEBSClient:~/oebs/gke$
ciuser@OpenEBSClient:~/oebs/gke$
ciuser@OpenEBSClient:~/oebs/gke$ terraform plan
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.


------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  + google_container_cluster.primary
      id:                                    <computed>
      additional_zones.#:                    <computed>
      addons_config.#:                       <computed>
      cluster_ipv4_cidr:                     <computed>
      enable_kubernetes_alpha:               "false"
      enable_legacy_abac:                    "true"
      endpoint:                              <computed>
      initial_node_count:                    "2"
      instance_group_urls.#:                 <computed>
      logging_service:                       <computed>
      master_auth.#:                         "1"
      master_auth.0.client_certificate:      <computed>
      master_auth.0.client_key:              <computed>
      master_auth.0.cluster_ca_certificate:  <computed>
      master_auth.0.password:                <sensitive>
      master_auth.0.username:                "admin"
      master_version:                        <computed>
      monitoring_service:                    <computed>
      name:                                  "oebs-staging"
      network:                               "default"
      network_policy.#:                      <computed>
      node_config.#:                         "1"
      node_config.0.disk_size_gb:            <computed>
      node_config.0.image_type:              "Ubuntu"
      node_config.0.local_ssd_count:         <computed>
      node_config.0.machine_type:            "n1-standard-2"
      node_config.0.oauth_scopes.#:          "4"
      node_config.0.oauth_scopes.1277378754: "https://www.googleapis.com/auth/monitoring"
      node_config.0.oauth_scopes.1632638332: "https://www.googleapis.com/auth/devstorage.read_only"
      node_config.0.oauth_scopes.172152165:  "https://www.googleapis.com/auth/logging.write"
      node_config.0.oauth_scopes.299962681:  "https://www.googleapis.com/auth/compute"
      node_config.0.preemptible:             "false"
      node_config.0.service_account:         <computed>
      node_pool.#:                           <computed>
      node_version:                          <computed>
      project:                               <computed>
      zone:                                  "us-central1-a"


Plan: 1 to add, 0 to change, 0 to destroy.

------------------------------------------------------------------------

Note: You didn't specify an "-out" parameter to save this plan, so Terraform
can't guarantee that exactly these actions will be performed if
"terraform apply" is subsequently run.

ciuser@OpenEBSClient:~/oebs/gke$ terraform apply

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  + google_container_cluster.primary
      id:                                    <computed>
      additional_zones.#:                    <computed>
      addons_config.#:                       <computed>
      cluster_ipv4_cidr:                     <computed>
      enable_kubernetes_alpha:               "false"
      enable_legacy_abac:                    "true"
      endpoint:                              <computed>
      initial_node_count:                    "2"
      instance_group_urls.#:                 <computed>
      logging_service:                       <computed>
      master_auth.#:                         "1"
      master_auth.0.client_certificate:      <computed>
      master_auth.0.client_key:              <computed>
      master_auth.0.cluster_ca_certificate:  <computed>
      master_auth.0.password:                <sensitive>
      master_auth.0.username:                "admin"
      master_version:                        <computed>
      monitoring_service:                    <computed>
      name:                                  "oebs-staging"
      network:                               "default"
      network_policy.#:                      <computed>
      node_config.#:                         "1"
      node_config.0.disk_size_gb:            <computed>
      node_config.0.image_type:              "Ubuntu"
      node_config.0.local_ssd_count:         <computed>
      node_config.0.machine_type:            "n1-standard-2"
      node_config.0.oauth_scopes.#:          "4"
      node_config.0.oauth_scopes.1277378754: "https://www.googleapis.com/auth/monitoring"
      node_config.0.oauth_scopes.1632638332: "https://www.googleapis.com/auth/devstorage.read_only"
      node_config.0.oauth_scopes.172152165:  "https://www.googleapis.com/auth/logging.write"
      node_config.0.oauth_scopes.299962681:  "https://www.googleapis.com/auth/compute"
      node_config.0.preemptible:             "false"
      node_config.0.service_account:         <computed>
      node_pool.#:                           <computed>
      node_version:                          <computed>
      project:                               <computed>
      zone:                                  "us-central1-a"


Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

google_container_cluster.primary: Creating...
  additional_zones.#:                    "" => "<computed>"
  addons_config.#:                       "" => "<computed>"
  cluster_ipv4_cidr:                     "" => "<computed>"
  enable_kubernetes_alpha:               "" => "false"
  enable_legacy_abac:                    "" => "true"
  endpoint:                              "" => "<computed>"
  initial_node_count:                    "" => "2"
  instance_group_urls.#:                 "" => "<computed>"
  logging_service:                       "" => "<computed>"
  master_auth.#:                         "" => "1"
  master_auth.0.client_certificate:      "" => "<computed>"
  master_auth.0.client_key:              "<sensitive>" => "<sensitive>"
  master_auth.0.cluster_ca_certificate:  "" => "<computed>"
  master_auth.0.password:                "<sensitive>" => "<sensitive>"
  master_auth.0.username:                "" => "admin"
  master_version:                        "" => "<computed>"
  monitoring_service:                    "" => "<computed>"
  name:                                  "" => "oebs-staging"
  network:                               "" => "default"
  network_policy.#:                      "" => "<computed>"
  node_config.#:                         "" => "1"
  node_config.0.disk_size_gb:            "" => "<computed>"
  node_config.0.image_type:              "" => "Ubuntu"
  node_config.0.local_ssd_count:         "" => "<computed>"
  node_config.0.machine_type:            "" => "n1-standard-2"
  node_config.0.oauth_scopes.#:          "" => "4"
  node_config.0.oauth_scopes.1277378754: "" => "https://www.googleapis.com/auth/monitoring"
  node_config.0.oauth_scopes.1632638332: "" => "https://www.googleapis.com/auth/devstorage.read_only"
  node_config.0.oauth_scopes.172152165:  "" => "https://www.googleapis.com/auth/logging.write"
  node_config.0.oauth_scopes.299962681:  "" => "https://www.googleapis.com/auth/compute"
  node_config.0.preemptible:             "" => "false"
  node_config.0.service_account:         "" => "<computed>"
  node_pool.#:                           "" => "<computed>"
  node_version:                          "" => "<computed>"
  project:                               "" => "<computed>"
  zone:                                  "" => "us-central1-a"
google_container_cluster.primary: Still creating... (10s elapsed)
google_container_cluster.primary: Still creating... (20s elapsed)
google_container_cluster.primary: Still creating... (30s elapsed)
google_container_cluster.primary: Still creating... (40s elapsed)
google_container_cluster.primary: Still creating... (50s elapsed)
google_container_cluster.primary: Still creating... (1m0s elapsed)
google_container_cluster.primary: Still creating... (1m10s elapsed)
google_container_cluster.primary: Still creating... (1m20s elapsed)
google_container_cluster.primary: Still creating... (1m30s elapsed)
google_container_cluster.primary: Still creating... (1m40s elapsed)
google_container_cluster.primary: Still creating... (1m50s elapsed)
google_container_cluster.primary: Still creating... (2m0s elapsed)
google_container_cluster.primary: Still creating... (2m10s elapsed)
google_container_cluster.primary: Still creating... (2m20s elapsed)
google_container_cluster.primary: Creation complete after 2m29s (ID: oebs-staging)

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
```

The screenshot of cluster created using terraform is here : 

image

File: e2e/terraform/gke/variables.tf [hunk 1]
```diff
@@ -0,0 +1,28 @@
+variable "credentials" {
+  default = "account.json"
+}
+
+variable "region" {
+  default = "us-central1-a"
+}
+
+variable "project" {
+  default = "strong-eon-153112"
```
Comment (kmova) [inline e2e/terraform/gke/variables.tf]: It may be a good idea to use place-holder names here? or ENV?

File: e2e/terraform/gke/variables.tf [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+variable "clustername" {
+  default = "oebs-staging"
+}
+
+variable "nodecount" {
```
Comment (AmitKumarDas) [inline e2e/terraform/gke/variables.tf]: Can we default to 3 nodes? Since replica count is defaulting to 3 now?

File: e2e/terraform/gke/README.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
+# STEPS TO CREATE GKE CLUSTER USING TERRAFORM
+
+## Step-1 : Obtain the Google Credentials from Google Developers Console 
+
+- Log into the Google Developers Console and select a project.
```
Comment (AmitKumarDas) [inline e2e/terraform/gke/README.md]: Will it be good to attach screen shots?

File: e2e/terraform/gke/README.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+- The API Manager view should be selected, click on "Credentials" on the left, then "Create credentials," and finally "Service account key."
+- Select "Compute Engine default service account" in the "Service account" dropdown, and select "JSON" as the key type.
```
Comment (AmitKumarDas) [inline e2e/terraform/gke/README.md]: Why `default` service account?

File: e2e/terraform/gke/README.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+- Clicking "Create" will download your credentials.
+- Rename this file to account.json and place in the directory where the terraform templates are present/will-be-placed
```
Comment (AmitKumarDas) [inline e2e/terraform/gke/README.md]: Can a screenshot be provided that shows the list of files in the particular directory?

File: e2e/terraform/gke/README.md [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+## Step-2 : Obtain the Google project ID
+
+- Log into the Google Developers Console to be sent to the Google API library page screen
+- Click the Select a project drop-down in the upper left
+- Copy the Project ID for the desired project from the window that shows
```
Comment (AmitKumarDas) [inline e2e/terraform/gke/README.md]: Where is the Project ID used later?

File: e2e/terraform/gke/oebs-staging.tf [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+resource "google_container_cluster" "primary" {
+  name = "${var.clustername}"
+  zone = "${var.region}"
+  initial_node_count = "${var.nodecount}"
+
+  master_auth {
+    username = "admin"
+    password = "oebsstaging@1234"
```
Comment (AmitKumarDas) [inline e2e/terraform/gke/oebs-staging.tf]: PASSWORD?

File: e2e/terraform/gke/oebs-staging.tf [hunk 8]
```diff
@@ -0,0 +1,29 @@
+
+provider "google" {
```
Comment (AmitKumarDas) [inline e2e/terraform/gke/oebs-staging.tf]: The code is the place where we should strive to provide maximum comments that lets others understand the essentials. In short, can we have good comments?

File: e2e/terraform/gke/variables.tf [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+  default = "2"
+}
+
+variable "machinetype" {
```
Comment (AmitKumarDas) [inline e2e/terraform/gke/variables.tf]: Comments explaining the characteristics of typical machinetypes?

File: e2e/terraform/gke/oebs-staging.tf (hunk 7 above)
Comment (ksatchit) [reply e2e/terraform/gke/oebs-staging.tf]: This is a placeholder @AmitKumarDas . However, will change it to an ENV based on review

File: e2e/terraform/gke/variables.tf (hunk 1 above)
Comment (ksatchit) [reply e2e/terraform/gke/variables.tf]: yes @kmova , will use ENV

File: e2e/terraform/gke/README.md (hunk 4 above)
Comment (ksatchit) [reply e2e/terraform/gke/README.md]: It has the necessary permissions to create the cluster resources. We can use other sa with adequate permissions too (added a note in the README)

File: e2e/terraform/gke/README.md (hunk 6 above)
Comment (ksatchit) [reply e2e/terraform/gke/README.md]: We will use it in the variables.tf file. Added a step in the README alluding to this.

File: e2e/terraform/gke/oebs-staging.tf (hunk 8 above)
Comment (ksatchit) [reply e2e/terraform/gke/oebs-staging.tf]: Sure. Added comments @AmitKumarDas

File: e2e/terraform/gke/README.md (hunk 4 above)
Comment (ksatchit) [reply e2e/terraform/gke/README.md]: Removed the placeholders to use TF_VAR environment variables.

**Pull request merged**